geode-core/src/main/java/org/apache/geode/internal/cache/backup/BackupDefinition.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.internal.cache.backup;
+
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.geode.cache.DiskStore;
+
+class BackupDefinition {
```

Comment:
Since BackupDefinition is passed to BackupDestination it seems like it would be better to pass an interface instead of a class. If someone implements their own BackupDestination I think they would only want the BackupDefinition interface to have gettor methods. They are not interested in creating a definition but in simply reading an existing definition passed to their implementation of the backupFiles method.

Why is RestoreScript part of the BackupDefinition? From an implementor of backupFiles it seems like it would either want to know the "Path" that is the restore script. But it is unclear how the restore script will work for other destinations. Isn't all the meta info the restore script needs encoded already in the BackupDefinition itself (ie. all the info on different types of files).

Why is the definition so complex with all the different types of files. For the purposes of BackupDestination.backupFiles isn't all it needs is a list of Path instances to backup?

It is also unclear to me why this class uses "Path" instead of "File" for the files that are being backed up. Given either one you can create the other. Path is an interface and File is a class so I would vote for Path. Just wondering if other reasons exist for using it.

Reply:
Good point on using an interface that only exposes the getters. 

The RestoreScript is part of the BackupDefinition to remove its logic from the BackupManager. Ultimately, we want to no longer have the script and instead have a manifest file that details the files and checks that need to be done to restore the backup (and do so in gemfire, not in a command line script), but that is a future feature.

The different file types are because we store those files in different directories in the backup, to prevent name collisions among other things. It is true that we could probably just have a list of files and have the methods that handle each type of file modify the path in the correct way. I think that the logic of where in the backup the file should go does belong here instead of in BackupManager.

I use Paths, generally, because they are now the prefered method of handling file IO (see https://docs.oracle.com/javase/tutorial/essential/io/legacy.html#), but also the Files utility class takes Paths only (added at the same time as Path) and I generally prefer using it (though there are still a few things that are easier to do with Apache Commons)

Reply:
But what would an implementor of BackupDestination do with the RestoreScript? Are they responsible for copying it? It seems like it might fit better on FileSystemBackupDestination

Reply:
Since we do not yet have a mechanism for getting backups returned to a system or restored to the correct locations, we still need that restore script. Once we have a way to restore the files without the script, we should be able to remove it in its entirety.